cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+    const auto diff = cmp(r1.start()->value(), r2.end()->value());
+    if (diff < 0) { // r1 start is strictly before r2 end.
+        return true;
+    }
+    if (diff > 0) { // r1 start is strictly after r2 end.
+        return false;
+    }
+    const auto len1 = r1.start()->value().representation().size();
+    const auto len2 = r2.end()->value().representation().size();
+    if (len1 == len2) { // The values truly are equal.
+        return r1.start()->is_inclusive() && r2.end()->is_inclusive();
+    } else if (len1 < len2) { // r1 start is a prefix of r2 end.
+        // a>=(1) starts before (a,b)<=(1,1) ends, but (a)>(1) doesn't.
+        return r1.start()->is_inclusive();
+    } else { // r2 end is a prefix of r1 start.
+        // (a,b)>=(1,1) starts before (a)<=(1) ends but after (a)<(1) ends.
+        return r2.end()->is_inclusive();
+    }
+}
+
+/// True if r1 end is before r2 end.
+bool ends_before_end(
+        const query::clustering_range& r1,
+        const query::clustering_range& r2,
+        const clustering_key_prefix::prefix_equal_tri_compare& cmp) {
+    if (!r1.end()) {
+        return false; // r1 end is +inf, which is after everything.
+    }
+    if (!r2.end()) {
+        return true; // r2 end is +inf, while r1 end is finite.
+    }
+    const auto diff = cmp(r1.end()->value(), r2.end()->value());
+    if (diff < 0) { // r1 end is strictly before r2 end.
+        return true;
+    }
+    if (diff > 0) { // r1 end is strictly after r2 end.
+        return false;
+    }
+    const auto len1 = r1.end()->value().representation().size();
+    const auto len2 = r2.end()->value().representation().size();
+    if (len1 == len2) { // The values truly are equal.
+        return !r1.end()->is_inclusive() || r2.end()->is_inclusive();
+    } else if (len1 < len2) { // r1 end is a prefix of r2 end.
+        // (a)<(1) ends before (a,b)<=(1,1), but (a)<=(1) doesn't.
+        return !r1.end()->is_inclusive();
+    } else { // r2 end is a prefix of r1 end.
+        // (a,b)<=(1,1) ends before (a)<=(1) but after (a)<(1).
+        return r2.end()->is_inclusive();
+    }
+}
+
+/// Correct clustering_range intersection.  See #8157.
+std::optional<query::clustering_range> intersection(
+        const query::clustering_range& r1,
+        const query::clustering_range& r2,
+        const clustering_key_prefix::prefix_equal_tri_compare& cmp) {
+    // Assume r1's start is to the left of r2's start.
+    if (starts_before_start(r2, r1, cmp)) {
+        return intersection(r2, r1, cmp);
+    }
+    if (!starts_before_end(r2, r1, cmp)) {
+        return {};
+    }
+    const auto& intersection_start = r2.start();
+    const auto& intersection_end = ends_before_end(r1, r2, cmp) ? r1.end() : r2.end();
+    if (intersection_start == intersection_end && intersection_end.has_value()) {
+        return query::clustering_range::make_singular(intersection_end->value());
+    }
+    return query::clustering_range(intersection_start, intersection_end);
```

Comment:
I'm disappointed we can't use interval's operations for that (but prefixes make that impossible without lots of work).

Reply:
What's stopping us from generalizing this code here to any prefix_compound_wrapper?